core/src/main/scala/org/http4s/Message.scala

```diff
@@ -211,6 +213,55 @@ sealed trait Message[F[_]] extends Media[F] { self =>
   def isChunked: Boolean =
     headers.get[`Transfer-Encoding`].exists(_.values.contains_(TransferCoding.chunked))
 
+  /** Puts a `Content-Length` header, replacing any existing.  Removes
+    * any existing `chunked` value from the `Transfer-Encoding`
+    * header.  It is critical that the supplied content length
+    * accurately describe the length of the body stream.
+    *
+    * {{{
+    * scala> import org.http4s.headers._
+    * scala> val chunked = Request().withHeaders(
+    *      |   `Transfer-Encoding`(TransferCoding.chunked),
+    *      |   `Content-Type`(MediaType.text.plain))
+    * scala> chunked.withContentLength(`Content-Length`.unsafeFromLong(1024)).headers
+    * res0: Headers = Headers(Content-Type: text/plain, Content-Length: 1024)
+    *
+    * scala> val chunkedGzipped = Request().withHeaders(
+    *      |   `Transfer-Encoding`(TransferCoding.chunked, TransferCoding.gzip),
+    *      |   `Content-Type`(MediaType.text.plain))
+    * scala> chunkedGzipped.withContentLength(`Content-Length`.unsafeFromLong(1024)).headers
+    * res1: Headers = Headers(Transfer-Encoding: gzip, Content-Type: text/plain, Content-Length: 1024)
+    *
+    * scala> val const = Request().withHeaders(
+    *      |   `Content-Length`(2048),
+    *      |   `Content-Type`(MediaType.text.plain))
+    * scala> const.withContentLength(`Content-Length`.unsafeFromLong(1024)).headers
+    * res1: Headers = Headers(Content-Type: text/plain, Content-Length: 1024)
+    * }}}
+    */
+  def withContentLength(contentLength: `Content-Length`): Self =
+    transformHeaders(_.transform { hs =>
+      val b = List.newBuilder[Header.Raw]
+      hs.foreach { h =>
+        h.name match {
+          case `Transfer-Encoding`.name =>
+            `Transfer-Encoding`
```

Comment:
I feel that we could do here impure but faster things like: 
```scala
if (h.value.contains(","))
  TransferCoding
    .parseList(h.value)
    .fold(
      _ => b += h,
      _.toList.foreach {
        case TransferCoding.chunked =>
          ()
        case tc =>
          b += Header.Raw(`Transfer-Encoding`.name, tc.coding)
      },
    )
else
  TransferCoding
    .parse(h.value)
    .fold(
      _ => b += h,
      {
        case TransferCoding.chunked =>
          ()
        case tc =>
          b += Header.Raw(`Transfer-Encoding`.name, tc.coding)
      },
    )
```
On one hand, this exposes the internals of the `Transfer-Encoding` header. One could fix that by placing that function within `Transfer-Encoding` or  `TransferCoding` and making it private. On another, in the case of separated `Transfer-Encoding` headers, we will have a performance boost. But this could be a kinda premature optimization. I haven't numbers for proof.

Reply:
Also, I wonder - are we really should parse `TransferCoding` values. Or we could rely on existing header values. If so, then we can remove the parsing step in case of a single `TransferCoding` value, I guess.

Reply:
Yeah, this strikes me as a premature optimization, but maybe we could do a separate PR if it's thought to be worthwhile.